# Dataset Profiling – thelook_ecommerce

This notebook documents a reproducible process to discover and validate the schema and table structure of `bigquery-public-data.thelook_ecommerce` using BigQuery metadata views.

## Dataset Context

This notebook supports **schema and structural profiling** of the `thelook_ecommerce` dataset hosted in Google BigQuery.

At this stage, the full schema has **not yet been assumed or hard-coded**. Instead, the objective of this notebook is to:

- Discover available tables directly from the source dataset  
- Inspect column names and data types using system metadata  
- Validate table structures before downstream analysis  
- Ensure that documentation reflects the authoritative source schema  

All schema information captured in this notebook will be derived directly from:

`bigquery-public-data.thelook_ecommerce.INFORMATION_SCHEMA`



## Discovering Available Tables

The following query retrieves the list of all tables available in the dataset directly from BigQuery metadata.

```sql
SELECT table_name
FROM `bigquery-public-data.thelook_ecommerce.INFORMATION_SCHEMA.TABLES`
ORDER BY table_name;


## Discovering Schema (Columns and Data Types)

After identifying relevant tables, the following query retrieves column names and data types directly from BigQuery metadata.

```sql
SELECT
  table_name,
  column_name,
  data_type
FROM `bigquery-public-data.thelook_ecommerce.INFORMATION_SCHEMA.COLUMNS`
ORDER BY table_name, ordinal_position;


## Table Sizes (Row Counts)

To understand dataset scale, the following query provides row counts for each table:

```sql
SELECT
  table_name,
  row_count
FROM `bigquery-public-data.thelook_ecommerce.__TABLES__`
ORDER BY row_count DESC;


## Structural Observations (To be completed)

After running the above queries, this section will summarize key structural insights such as:

- Which tables are most relevant to transaction-level analysis  
- Which tables contain customer, product, and event-level data  
- Key identifier fields (e.g., primary keys and join fields)  
- Any notable schema constraints or limitations  

This section will be completed after reviewing query results.


## Captured Results (To be filled after running queries)

Paste or summarize the results from BigQuery here:

- **Tables discovered:** (paste list or key tables)
- **Row counts:** (paste top tables + counts)
- **Key join fields identified:** (e.g., order_id, user_id, product_id)
- **Notable fields for returns/profit erosion:** (e.g., status, returned_at, sale_price, cost)

This section provides evidence that schema profiling was performed using the authoritative source.

In [1]:
import pandas as pd

# Load raw datasets
orders = pd.read_csv("../data/raw/orders.csv")
order_items = pd.read_csv("../data/raw/order_items.csv")
products = pd.read_csv("../data/raw/products.csv")
users = pd.read_csv("../data/raw/users.csv")

datasets = {
    "orders": orders,
    "order_items": order_items,
    "products": products,
    "users": users
}

# Profiling output
for name, df in datasets.items():
    print(f"\n=== {name.upper()} ===")
    print("Rows:", df.shape[0])
    print("Columns:", df.shape[1])
    print("\nColumn names:")
    print(df.columns.tolist())
    print("\nData types:")
    print(df.dtypes)



=== ORDERS ===
Rows: 125017
Columns: 9

Column names:
['order_id', 'user_id', 'status', 'gender', 'created_at', 'returned_at', 'shipped_at', 'delivered_at', 'num_of_item']

Data types:
order_id         int64
user_id          int64
status          object
gender          object
created_at      object
returned_at     object
shipped_at      object
delivered_at    object
num_of_item      int64
dtype: object

=== ORDER_ITEMS ===
Rows: 181469
Columns: 11

Column names:
['id', 'order_id', 'user_id', 'product_id', 'inventory_item_id', 'status', 'created_at', 'shipped_at', 'delivered_at', 'returned_at', 'sale_price']

Data types:
id                     int64
order_id               int64
user_id                int64
product_id             int64
inventory_item_id      int64
status                object
created_at            object
shipped_at            object
delivered_at          object
returned_at           object
sale_price           float64
dtype: object

=== PRODUCTS ===
Rows: 29120
Columns:

C:\Users\avinm\AppData\Local\Temp\ipykernel_19732\2744468595.py:7: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  users = pd.read_csv("../data/raw/users.csv")


## Captured Profiling Results

Profiling was performed on the locally extracted raw CSV files originating from the BigQuery public dataset.

- **orders**: 125,017 rows, 9 columns  
- **order_items**: 181,469 rows, 11 columns  
- **products**: 29,120 rows, 9 columns  
- **users**: 100,000 rows, 16 columns  

### Key Observations

- Shared identifiers such as `order_id`, `user_id`, and `product_id` support relational joins across tables.  
- Time-related fields (`created_at`, `shipped_at`, `delivered_at`, `returned_at`) support lifecycle and return-timing analysis.  
- Pricing and value-related fields (`sale_price`, `retail_price`, `cost`) support margin and profit erosion analysis.  
- The presence of customer attributes (e.g., age, gender, location, traffic_source) enables segmentation and behavioral analysis.  

These results confirm that the dataset structure supports transaction-level analytical workflows and aligns with the objectives of this project.
